src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/JsonFileMerger.cs

```diff
@@ -0,0 +1,171 @@
+        var targetRepoCurrentJson = await GetJsonFromGit(targetRepo, targetRepoJsonRelativePath, targetRepoCurrentRef, allowMissingFiles);
+
+        var sourcePreviousJson = await GetJsonFromGit(sourceRepo, sourceRepoJsonRelativePath, sourceRepoPreviousRef, allowMissingFiles);
+        var sourceCurrentJson = await GetJsonFromGit(sourceRepo, sourceRepoJsonRelativePath, sourceRepoCurrentRef, allowMissingFiles);
+
+        if (!allowMissingFiles || !await DeleteFileIfRequiredAsync(
+                targetRepoPreviousJson,
+                targetRepoCurrentJson,
+                sourcePreviousJson,
+                sourceCurrentJson,
+                targetRepo.Path,
+                targetRepoJsonRelativePath,
+                targetRepoCurrentRef,
+                EmptyJsonString))
+        {
+            var targetRepoChanges = SimpleConfigJson.Parse(targetRepoPreviousJson).GetDiff(SimpleConfigJson.Parse(targetRepoCurrentJson));
+            var vmrChanges = SimpleConfigJson.Parse(sourcePreviousJson).GetDiff(SimpleConfigJson.Parse(sourceCurrentJson));
+
+            VersionFileChanges<JsonVersionProperty> mergedChanges = MergeChanges(
+                targetRepoChanges,
+                vmrChanges,
+                SelectNewerJsonVersionProperty,
+                targetRepoJsonRelativePath,
+                property => property.Replace(':', '.'));
+
+            var currentJson = await GetJsonFromGit(targetRepo, targetRepoJsonRelativePath, "HEAD", allowMissingFiles);
+            var mergedJson = SimpleConfigJson.ApplyJsonChanges(currentJson, mergedChanges);
+
+            var newJson = new GitFile(targetRepo.Path / targetRepoJsonRelativePath, mergedJson);
+
+            await _gitRepoFactory.CreateClient(targetRepo.Path)
+                .CommitFilesAsync(
+                    [newJson],
+                    targetRepo.Path,
+                    targetRepoCurrentRef,
+                    $"Merge {targetRepoJsonRelativePath} changes from VMR");
+        }
+
+        await targetRepo.StageAsync(["."]);
+    }
+
+    private static async Task<string> GetJsonFromGit(ILocalGitRepo repo, string jsonRelativePath, string reference, bool allowMissingFile) =>
+        reference == Constants.EmptyGitObject
+            ? EmptyJsonString
+            : await repo.GetFileFromGitAsync(jsonRelativePath, reference)
+                ?? (allowMissingFile
+                    ? EmptyJsonString
+                    : throw new FileNotFoundException($"File not found at {repo.Path / jsonRelativePath} for reference {reference}"));
+
+    private static JsonVersionProperty SelectNewerJsonVersionProperty(JsonVersionProperty repoProp, JsonVersionProperty vmrProp)
+    {
+        if (repoProp.Value == null && vmrProp.Value == null)
+        {
+            throw new ArgumentException($"Compared values for '{repoProp.Name}' are null");
```

Comment:
so you'll be unblocking these in a followup?

Reply:
Yes